hax/treemakers/posrec.py

```diff
@@ -0,0 +1,200 @@
+        loaded_model_json = json_file_nn.read()
+        json_file_nn.close()
+        loaded_nn_model = model_from_json(loaded_model_json)
+        weights_file = utils.data_file_name(hax.config['neural_network_weights'])
+        loaded_nn_model.load_weights(weights_file)
+        loaded_nn_model.compile(loss='mean_squared_error', optimizer='adam')
+        self.nn_tensorflow = loaded_nn_model
+        self.list_bad_pmts = [1, 2, 12, 26, 34, 62, 65, 79, 86, 88, 102, 118,
+                              130, 134, 135, 139, 148, 150, 152, 162, 178, 183,
+                              190, 198, 206, 213, 214, 234, 239, 244, 27, 73,
+                              91, 137, 167, 203]
+        self.ntop_pmts = 127  # How to get this automatically?
+                
+    def get_data(self, dataset, event_list=None):
+
+        data, _ = hax.minitrees.load_single_dataset(dataset, ['Corrections',
+                                                              'Fundamentals'])
+        self.x = data.x_3d_nn.values
+        self.y = data.y_3d_nn.values
+        self.z = data.z_3d_nn.values
+
+        self.indices = list(data.event_number.values)
+
+        return hax.minitrees.TreeMaker.get_data(self, dataset, event_list)
+
+    def extract_data(self, event):
+
+        event_data = {
+            "s1_pattern_fit_hax": None,
+            "s1_pattern_fit_hits_hax": None,
+            "s1_pattern_fit_bottom_hax": None,
+            "s1_pattern_fit_bottom_hits_hax": None,
+            "s1_area_fraction_top_probability_hax": None,
+            "x_observed_nn_tf": None,
+            "y_observed_nn_tf": None,
+            "s1_area_upper_injection_fraction": None,
+            "s1_area_lower_injection_fraction": None,
+        }
+
+        # We first need the positions. This minitree is only valid when loading
+        # Corrections since you need that to get the corrected positions
+        if not len(event.interactions):
+            return event_data
+        e = event.event_number
+        try:
+            i = self.indices.index(e)
+        except Exception:
+            return event_data
+        interaction = event.interactions[0]
+        s1 = event.peaks[interaction.s1]
+        s2 = event.peaks[interaction.s2]
+
+        # Position reconstruction based on NN from TensorFlow
+        s2apc = np.array(list(s2.area_per_channel))
+        s2apc_clean = []
+        for i, s2_t in enumerate(s2apc):
+            if i not in self.list_bad_pmts and i < self.ntop_pmts:
+                s2apc_clean.append(s2_t)
+        s2apc_clean = np.asarray(s2apc_clean)
+        s2apc_clean_norm = s2apc_clean / s2apc_clean.sum()
+        s2apc_clean_norm = s2apc_clean_norm.reshape(1, len(s2apc_clean_norm))
+        predicted_xy_tensorflow = self.nn_tensorflow.predict(s2apc_clean_norm)
+        event_data['x_observed_nn_tf'] = predicted_xy_tensorflow[0, 0] / 10.
+        event_data['y_observed_nn_tf'] = predicted_xy_tensorflow[0, 1] / 10.
```

Comment:
Can this preprocessing be broken into its own function?  If I am to add another model, then it's copy-paste 10 more lines... though the preprocessing is slightly different.

Also, clear is the wrong word.  It's 'preprocessed' or 'normalized'.

Reply:
@tunnell It looks like you have some nice ideas on how to clean up the codes here. Can you modify this PR and test it on a dataset. Maybe you can also fix the issue that sometimes TF fails to be running.